nomad/structs/structs.go

```diff
@@ -6460,12 +6460,20 @@ func (tg *TaskGroup) Canonicalize(job *Job) {
 
 // NomadServices returns a list of all group and task - level services in tg that
 // are making use of the nomad service provider.
+//
+// The returned services must be canonicalized and go though a taskenv.TaskEnv
+// to have its values interpolated.
 func (tg *TaskGroup) NomadServices() []*Service {
 	return tg.filterServices(func(s *Service) bool {
 		return s.Provider == ServiceProviderNomad
 	})
 }
 
+// ConsulServices returns a list of all group and task - level services in tg
+// that are making use of the consul service provider.
+//
+// The returned services must be canonicalized and go though a taskenv.TaskEnv
+// to have its values interpolated.
```

Comment:
I kind of want to force the caller to pass in the job name, namespace, and a `TaskEnv` so we can return services that are already interpolated, but maybe that's too much?

Reply:
After stewing on this a bit more I think I'm going to add the arguments after all, but allow passing empty values. It's like checking and signing a "I have read and agree to the Terms and Conditions" that you know the values returned are not interpolated 😅

Reply:
Small refactoring of service interpolation in https://github.com/hashicorp/nomad/pull/16413. I will rebase this PR when that one gets merged.

Reply:
Not refactoring here. There are quite a few intricacies that would require a more thoughtful process.